### DISCUSSION TITLE
How to use on_change templating with imported lists

### DISCUSSION THREAD

**Discussion by arrrgi** [twpayne/chezmoi#2916] [Q&A]
Borrowing from @woolsweater previous discussion, I'm trying to implement something similar by breaking out a list of brew packages (or whatever other package utility) into a separate template that is hashed for changes, but I'm getting an error when I try to test it.

I've chosen an approach to seperate the bash script template and provide data to the template from another, rather than baking all the lists and variables into the main template.

It seems I've misinterpreted the docs on how I can achieve this so that:

**.chezmoiscripts/run_onchange_after_00_install-packages.sh.tmpl**
```
#!/bin/env bash

set -eufo pipefail

{{- $script := joinPath .chezmoi.sourceDir ".chezmoitemplates/packages.tmpl" -}}
# hash: {{- include $script | sha256sum }}
{{- $pkgs := includeTemplate $script . -}}

brew bundle --no-lock --file=/dev/stdin << EOF || :
{{ range ($pkgs.taps | sortAlpha | uniq) -}}
tap "{{ . }}"
{{ end }}
{{ range ($pkgs.brews | sortAlpha | uniq) -}}
brew "{{ . }}"
{{ end }}
{{ range ($pkgs.casks | sortAlpha | uniq) -}}
cask "{{ . }}"
{{ end }}
EOF
```

takes an input from:

**.chezmoitemplates/packages.tmpl**
```
{{ $taps := list
  "1password/tap"
  "homebrew/bundle"
  "homebrew/cask"
-}}

{{ $brews := list
  "bat"
  "coreutils"
-}}

{{ $casks := list
  "vlc"
  "wezterm"
-}}
```

but executing this give me an error like so:

```
> chezmoi execute-template < .chezmoiscripts/run_onchange_after_00_install-packages.sh.tmpl
chezmoi: template: stdin:10:15: executing "stdin" at <$pkgs.taps>: can't evaluate field taps in type string
```

Any tips that would help me understand this much better and correct my error here would be greatly appreciated!

**Comment by twpayne** [comment]
`includeTemplate` returns a string, not structured data.

Variables like `$taps` are local to the template. They do not modify the globals of the template calling `includeTemplate`.

For what you are looking to achieve here, I would recommend either using [`.chezmoidata`](https://www.chezmoi.io/reference/special-files-and-directories/chezmoidata-format/), or having the included template return JSON and then parsing that JSON data (warning: untested code):

```
{{ $taps := list
  "1password/tap"
  "homebrew/bundle"
  "homebrew/cask"
-}}
{{ dict "taps" $taps | toJson }}
```

```
{{ $dict := includeTemplate "packages.tmpl" | fromJson }}
{{ get $dict "taps" }}
```

**Comment by arrrgi** [comment]
Thanks @twpayne, that makes sense now.

I ended up finding another solution to this by moving

```Go
{{ range ($taps | sortAlpha | uniq) -}}
tap "{{ . }}"
{{ end -}}
{{ range ($brews | sortAlpha | uniq) -}}
brew "{{ . }}"
{{ end -}}
{{ range ($casks | sortAlpha | uniq) -}}
cask "{{ . }}"
{{ end -}}
```
 into **packages.tmpl** so the variables were local to the sprig functions, resulting in the imported template already being the rendered Brewfile.

No I'm just stuck on how to build a dictionary for `mas {{$app}}, {{$id}}` from

```
{{ $masapps := `
  "Bear", id: 1091189122
  "Lacona", id: 1495926282
  "Moom", id: 419330170
  "Tailscale", id: 1475387142
  "Todoist", id: 585829637
`-}}
```
as dropping the string '`' results in an error about an expected ','

**Reply by twpayne** [reply]
For this you can use the `dict` function:

```
{{ $masapps := dict
  "Bear" 1091189122
  "Lacona" 1495926282
  "Moom" 419330170
  "Tailscale" 1475387142
  "Todoist" 585829637
-}}
```

This will give you a dict where the keys are `app`s and the values are `id`s.

**Comment by arrrgi** [comment]
Thanks, makes sense now!
